app/controllers/more-info.js

```diff
@@ -18,26 +26,111 @@ export default Ember.Controller.extend({
+    if (this.model.get('attribution')) {
+      return this.checkFormErrors(changeset, 'attribution_text')
+    } else {
+      if (!this.attributionSelected) return new Promise (resolve => resolve(['Attribution Error 1']));
+      else return new Promise (resolve => resolve([]));
+    }
+  },
+  checkFrequencyError: function () {
+    return new Promise (resolve => {
+      if (this.model.get('update_frequency')) resolve([]);
+      else resolve(['Frequency Error 1']);
+    })
+  },
+  checkErrors: function (changeset) {
+    let totalErrorMessages = [];
+    return this.checkLicenseError(changeset)
+      .then((licenseErrorMessages) => licenseErrorMessages.map((eMessage) => totalErrorMessages.push(eMessage)))
+      .then(() => this.checkAttributionError(changeset))
+      .then((attributionErrorMessages) => attributionErrorMessages.map((eMessage) => totalErrorMessages.push(eMessage)))
+      .then(() => this.checkFrequencyError())
+      .then((frequencyErrorMessages) => {
+        frequencyErrorMessages.map((eMessage) => totalErrorMessages.push(eMessage));
+        return totalErrorMessages;
+      })
+  },
+  resetErrorState: function () {
+    Ember.set(this, 'showErrorState', false);
+    Ember.set(this, 'errorMessages', []);
+  },
   actions: {
-    licenseExists: function(input){
+    licenseExists: function(input) {
+      // Check user interacted with attribution at all
+      this.licenseSelected = true;
       this.model.set('license_exists', input)
       this.model.set('license', null);
       this.model.set('license_url', null);
       this.model.set('user_submitted_url', null);
-      this.model.set('attribution', null);
-      this.model.set('attribution_text', null);
+      // Confirm this part
+      // this.model.set('attribution', null);
+      // this.model.set('attribution_text', null);
     },
     selectLicense: function(license){
       this.model.set('license', null)
       this.model.set('user_submitted_url', null)
       this.model.set('license', license.name)
       this.model.set('license_url', license.url)
     },
-    changeRoute: function(route){
-      this.transitionToRoute(route);
+    changeRoute: function(route, changeset){
+      this.checkErrors(changeset)
+        .then((errorMsgs) => {
+          // When there is any error, show it and do not proceed
+          if (errorMsgs.length) {
+            Ember.set(this, 'showErrorState', true);
+            Ember.set(this, 'errorMessages', errorMsgs);
+          // When there is no error message, proceed
+          } else {
+            this.resetErrorState();
+            this.transitionToRoute(route);
+          }
+        })
+        .catch ((err) => {
+          const errorMsgs = [err];
+          Ember.set(this, 'showErrorState', true);
+          Ember.set(this, 'errorMessages', errorMsgs);
+        })
     },
     userSubmittedLicense: function(input){
-      this.model.set('license', null)
+      // this.model.set('license', 'provide')
```

Comment:
This part was bit confusing for me. Does `license` need to be set as `null`? This gives rendering error when user started typing because input text for license only renders when license is `provide`

Reply:
@hanbyul-here yes, I believe `license` should be set to `null` when the user types the license url, because otherwise the `license` model attribute would retain the name selected from the dropdown, if the user did that before switching to enter their own url.

The input text for the user-submitted url only renders when the `licenseType' is `provide`.

Does this all sound ok to you?